ludwig/distributed/deepspeed.py

```diff
@@ -0,0 +1,183 @@
+import logging
+import os
+import warnings
+from typing import Any, Dict, Mapping, Optional, Tuple, TYPE_CHECKING
+
+import deepspeed
+import deepspeed.comm
+import torch
+from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_checkpoint
+from torch import nn
+from torch.optim.optimizer import Optimizer
+
+from ludwig.distributed.ddp import DDPStrategy
+from ludwig.modules.optimization_modules import get_optimizer_class_and_kwargs
+from ludwig.utils.checkpoint_utils import Checkpoint
+
+if TYPE_CHECKING:
+    from ludwig.modules.lr_scheduler import LRScheduler
+    from ludwig.schema.trainer import ECDTrainerConfig
+
+
+DEFAULT_ZERO_OPTIMIZATION = {
+    "stage": "auto",
+    "stage3_gather_16bit_weights_on_model_save": "auto",
+    "offload_optimizer": {"device": "auto"},
+    "offload_param": {"device": "auto"},
+}
+
+# Filter out warnings about DeepSpeed use of deprecated methods. Can remove on upgrade to DeepSpeed 0.9.
+warnings.filterwarnings(
+    action="ignore",
+    category=UserWarning,
+    module="torch.distributed.distributed_c10d",
+)
+
+
+class DeepSpeedStrategy(DDPStrategy):
+    def __init__(self, zero_optimization: Optional[Dict[str, Any]] = None, **kwargs):
+        # If we're initializing from a `deepspeed` CLI command, deepspeed will have already been initialized, as
+        # indicated by the presence of the LOCAL_RANK var. Otherwise, we're initializing from Ray / torchrun, and will
+        # need to set this var ourselves, then init DeepSpeed here.
+        local_rank, local_size = os.environ.get("LOCAL_RANK"), os.environ.get("LOCAL_SIZE")
+        init_deepspeed = local_rank is None or local_size is None
```

Comment:
If both of these are set, is the reason we don't need to call `deepspeed.init_distributed()` because deepspeed is already initialized? When would that be the case?

Reply:
We don't need to `init_distributed` when using Ray because we're already in a `torch.distributed` environment, but when using the `deepspeed` launch we need to call `init_distributed` to be in that context.

Reply:
Ahhh got it, nice!